src/Build/Graph/ProjectGraph.cs

```diff
@@ -29,6 +37,6 @@ internal sealed class ProjectGraph
         /// <summary>
         /// Get an unordered collection of all project nodes in the graph.
         /// </summary>
-        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes => Array.Empty<ProjectGraphNode>();
+        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes => _projectNodes;
```

Comment:
I'm starting to wonder if we need this API. It feels a bit weird to expose "here's a list of nodes; don't ever `foreach` over it" instead of graph traversal operations and maybe things like `NodesWithNoProjectReferences`. But maybe those really do belong in the caller? No need to change for this PR but let's keep thinking about this.

Reply:
I do think it's quite common still to get a collection of all unique projects. `foreach` is fine as long as it's not `foreach` -> build

Reply:
I guess people will use the nodes for whatever reason, including hand made builds. Unless we add explicit checks, it will happen :)